### PR TITLE
`Adaptive learning`: Fix prerequisites not shown in course enrollment

### PR DESCRIPTION
**Pull request by JohannesStoehr**
#### General

- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).

#### Server
- [x] **Important**: I implemented the changes with a [very good performance](https://docs.artemis.cit.tum.de/dev/guidelines/performance/) and prevented too many (unnecessary) and too complex database calls.
- [x] I **strictly** followed the principle of **data economy** for all database calls.
- [x] I **strictly** followed the [server coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I added pre-authorization annotations according to the [guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/#rest-endpoint-best-practices-for-authorization) and checked the course groups for all new REST Calls (security).

#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data (e.g. using paging).
- [x] I **strictly** followed the principle of **data economy** for all client-server REST calls.
- [x] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).

### Motivation and Context

Users got a not authorised exception when trying to see the prerequisites of a course before enrolling.

### Description

Changed the authorisation to see prerequisites of a course to all students regardless if they are already enrolled in the course.
Removed the competency rings when viewing the prerequisites before enrolling in a course, since the student does not yet have progress in them.

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Student not enrolled in the course

1. Log in to Artemis as instructor
2. Create a prerequisite in the course
3. Enable self enrolment in the course settings
4. Try to enrol in the course as student and view the prerequisites

[](https://artemis-test1.artemis.cit.tum.de)
[](https://artemis-test2.artemis.cit.tum.de)
[](https://artemis-test3.artemis.cit.tum.de)
[](https://artemis-test4.artemis.cit.tum.de)
[](https://artemis-test5.artemis.cit.tum.de)
[](https://artemis-test6.artemis.cit.tum.de)
[](https://artemis-test9.artemis.cit.tum.de)

#### Performance Review

- [x] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance even for very large courses with more than 2000 students.
- [x] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance even for very large courses with more than 2000 students.
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2
#### Performance Tests
- [x] Test 1
- [x] Test 2

#### Client

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| competency-card.component.ts | 100% | ✅ |

#### Server

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| PrerequisiteResource.java | 96% | ✅ |

## Summary by CodeRabbit

- **New Features**
	- Introduced a new input property `noProgressRings` to the competency card, allowing for conditional rendering of progress rings based on user context.
	- Updated competency card components to utilize the new input property, enhancing display logic in various contexts.

- **Bug Fixes**
	- Modified access control for the `getPrerequisitesWithProgress` method, broadening user roles that can access it.

- **Tests**
	- Renamed and updated test methods to reflect a shift in focus from student-specific competencies to course-related competencies.
	- Standardized input assignment in tests for the competency card component to improve consistency.

### CODE REVIEW

**Review by Strohgelaender — DISMISSED** [review]
code

**Review by Jan-Thurner — DISMISSED** [review]
Code LGTM

**Review by az108 — APPROVED** [review]
Tested on TS2, works as expected
image

**Review by coolchock — DISMISSED** [review]
tested on ts1 in the testing session, works as expected

File: src/main/webapp/app/course/competencies/competency-card/competency-card.component.html
Comment (Strohgelaender) [inline src/main/webapp/app/course/competencies/competency-card/competency-card.component.html:65]: This is nit-picky, but can we avoid the double negation by having a `showProgressRings` input that is true by default?

**Review by cremertim — APPROVED** [review]
Code changes lgtm. Thanks for incorporating signals

**Review by coolchock — DISMISSED** [review]
re-approve

**Review by BBesrour — APPROVED** [review]

**Review by az108 — DISMISSED** [review]
Tested on TS1 again after changes, still works as expected 👍 

image

File: src/main/webapp/app/course/competencies/competency-card/competency-card.component.html
Comment (JohannesStoehr) [reply src/main/webapp/app/course/competencies/competency-card/competency-card.component.html:65]: I thought about this as well, but at least I find it more intuitive to set `noProgressRings` to `true` to mark an exception in the normal behavior...

**Review by SimonEntholzer — APPROVED** [review]
Viewing Prerequisites during enrolment works nicly, only one small nit pick: the table for it looks a bit off in dark mode, as for some reason there is no line above the two buttons:
image

**Review by sarpsahinalp — DISMISSED** [review]
Code lgtm

**Comment by MaximilianAnzinger** [comment]
Please complete the PR description. The coverage table is missing and no all check boxes have been ticked.

**Review by Jan-Thurner — DISMISSED** [review]
new changes to the code lgtm

**Review by Jan-Thurner — APPROVED** [review]
reapprove after merge conflict

**Review by cremertim — APPROVED** [review]

**Review by BBesrour — APPROVED** [review]

**Review by JohannesWt — APPROVED** [review]
Tested after merge conflicts works 👍

**Review by MaximilianAnzinger — APPROVED** [review]
Maintainer approved

**Pull request merged**
